gadopt/transport_solver.py

```diff
@@ -390,3 +391,112 @@ def set_equation(self) -> None:
             approximation=self.approximation,
             bcs=self.weak_bcs,
         )
+
+
+class DiffusiveSmoothingSolver(GenericTransportSolver):
+    """A class to perform diffusive smoothing by inheriting from GenericTransportSolver.
+
+    This class provides functionality to solve a diffusion equation for smoothing
+    a scalar function, using clean inheritance from GenericTransportSolver.
+
+    Args:
+        function_space (firedrake.FunctionSpace): The function space to solve the diffusion equation.
+        wavelength (Number): The wavelength for diffusion.
+        K (Union[firedrake.Function, Number], optional): Diffusion tensor. Defaults to 1.
+        bcs (dict[int, dict[str, int | float]] | None): Boundary conditions to impose on the solution.
+        solver_parameters (dict[str, str | float] | None): Solver parameters for the solver. Defaults to None.
+        integration_quad_degree (int | None): Quadrature degree for integrating the diffusion tensor. If None, defaults to 2p+1 where p is the polynomial degree.
+        **kwargs: Additional keyword arguments to pass to the solver.
+    """
+
+    def __init__(
+        self,
+        function_space: FunctionSpace,
+        wavelength: Number,
+        K: Function | Number = 1,
+        bcs: dict[int, dict[str, int | float]] | None = None,
+        solver_parameters: dict[str, str | float] | None = None,
+        integration_quad_degree: int | None = None,
+        **kwargs
+    ):
+        # Create solution function
+        solution = Function(function_space, name="smoothed")
+
+        # Calculate diffusive time step
+        dt = self._calculate_diffusive_time_step(function_space, wavelength, K, integration_quad_degree)
+
+        # Convert boundary conditions to the format expected by GenericTransportSolver
+        converted_bcs = self._convert_bcs(bcs or {})
+
+        # Initialize the parent GenericTransportSolver
+        super().__init__(
+            "diffusion",
+            solution,
+            dt,
+            BackwardEuler,
+            eq_attrs={"diffusivity": ensure_constant(K)},
+            bcs=converted_bcs,
+            solver_parameters=solver_parameters,
+            **kwargs
+        )
+
+    def _convert_bcs(self, bcs: dict[int, dict[str, int | float]]) -> dict[int, dict[str, int | float]]:
+        """Convert boundary conditions to the format expected by GenericTransportSolver."""
+        transport_bcs = {}
+
+        for subdomain_id, bc in bcs.items():
+            for bc_type, value in bc.items():
+                if bc_type == 'T':
+                    # GenericTransportSolver expects "g" for Dirichlet boundary conditions
+                    transport_bcs[subdomain_id] = {"g": value}
+                else:
+                    raise ValueError("Boundary conditions other than Dirichlet are not supported.")
+
+        return transport_bcs
```

Comment:
Do we need this method? Smoothing should be applicable to any field, and it is not clear to me why the strong bcs tag should be `T`. If `g` (for general) is not suitable, we could use `s` (for smoothing). Additionally, weak bcs are allowed in `diffusion_term`, so if people, for some reason, want to use them, shouldn't they be allowed to?